File: october/week4/search_in_a_sorted_array_of_unknown_size.py
# """
# This is ArrayReader's API interface.
# You should not implement it, or speculate about its implementation
# """
#class ArrayReader:
#    def get(self, index: int) -> int:

class Solution:
    def __init__(self):
        super().__init__()
        self.delim_out = 2147483647

    def search(self, reader, target):
        """
        :type reader: ArrayReader
        :type target: int
        :rtype: int
        """
        st = 0
        ed = 10000
        while st <= ed:
            mid = (st+ed) // 2
            cur = reader.get(mid)
            if cur == target:
                return mid

            elif cur < target:
                st = mid + 1
                
            else:
                ed = mid - 1

        return -1 

File: october/week4/search_in_a_sorted_array_of_unknown_size.cpp
/**
 * // This is the ArrayReader's API interface.
 * // You should not implement it, or speculate about its implementation
 * class ArrayReader {
 *   public:
 *     int get(int index);
 * };
 */

class Solution {
public:
    int search(const ArrayReader& reader, int target) {
        int st=0, ed=10000;
        while(st<=ed){
            int mid = (st+ed)/2;
            int cur = reader.get(mid);
            if(target == cur)   return mid;
            else if(target < cur)   ed = mid -1;
            else    st = mid + 1;
        }
        return -1;
    }
};
